pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -734,10 +734,18 @@ func getNodeOrPVCName(d *appsv1.Deployment) (string, error) {
 		return v, nil // OSD is on PVC
 	}
 	for k, v := range d.Spec.Template.Spec.NodeSelector {
-		if k == corev1.LabelHostname {
+		if k == k8sutil.LabelHostname() {
 			return v, nil
 		}
 	}
+	// try to fallback on previous hostname label
+	// NodeSelector always has a single entry
+	if len(d.Spec.Template.Spec.NodeSelector) == 1 {
+		for _, v := range d.Spec.Template.Spec.NodeSelector {
+			return v, nil
+		}
+	}
+
```

Comment:
Do we think fallback is going to be needed here? I don't see fallback in other places, so I'm kind of curious why here. 

I am also a bit concerned about falling back at all here. I have seen that user bugs tend to come up in corner cases when Rook chooses to fallback instead of throw an error if something is unclear. 

Fallback isn't bad, but I think it would be good to make sure we've discussed (and maybe commented) why we think it's useful and safe here.

Reply:
@BlaineEXE deployment in this function is existing deployment containing previous hostname label.
Without this fallback osd nodeSelector won't be updated to a new label for exsiting cluser. On [line 737](https://github.com/rook/rook/pull/15255/files/25ee6b4f59558f63e6ab05e53b2b45c81fa88285#diff-813ce9bdffbcd2753357c7f27d6436327e55395a0d8e9962172161159feef02aR737)  `k="kubernetes.io/hostname"` and `k8sutil.LabelHostname()="new_host_label"`

Reply:
I got it now. This is reading deployed resources, which might be an old version/config. This isn't directly setting what is being deployed. Thanks for clarifying